### ISSUE TITLE
Paste slow

### ISSUE THREAD

**Issue by Mellbourn** [microsoft/vscode#235959]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.96.0
- OS Version: macOS 14.7.1

Steps to Reproduce:

1. When I paste into a TypeScript file, it is very slow. It shows a small circular progress indicator for several seconds before finishing the paste

This is a very large repo, but pasting was fast even in this repo before version 1.96.

**Comment by mjbvz** [comment]
This is very likely https://code.visualstudio.com/updates/v1_96#_paste-with-imports-for-javascript-and-typescript. As a workaround you can set `"typescript.updateImportsOnPaste.enabled": false` / `"javascript.updateImportsOnPaste.enabled": false`

To help us investigate the perf issue, are you able to share the project? If not, can you try collecting the TypeScript server logs:

1. Set `"typescript.tsserver.log": "verbose",`
1. Restart VS Code and reproduce the problem
1. In VS Code, run the `TypeScript: Open TS Server log` command
1. This should open a large log file called `tsserver.log` 

ThIs should let us know how long paste is taking

 > **⚠️Warning:** The TypeScript log may include information from your workspace, including file paths and source code. If you have any concerns about posting this publicly on Github, just let me know and we can arrange something else. On our side, we only use these logs to investigate issues like this

**Comment by Mellbourn** [comment]
Yes, setting `"typescript.updateImportsOnPaste.enabled": false` fixed the issue immediately.

I have reproduced the problem as you suggested (and without extensions). Pasting took more than half a minute :).  I have a logfile (81MB). 
I cannot share the file publicly. I might be able to share it privately.

**Comment by imarkc** [comment]
We are experiencing this issue as well. It broke intellisense for large projects using TypeScript.

`"typescript.updateImportsOnPaste.enabled": false` only fixed the pasting issue

**Comment by JavaScriptBach** [comment]
@mjbvz I just wanted to check whether you still need a tsserver log? I can't share my tsserver.log as-is, but I may be able to share a sanitized one that anonymizes file paths + import specifiers, if that is still helpful.

I'd also love to understand the expected behavior of this feature better. My understanding from reading the [release notes](https://code.visualstudio.com/updates/v1_96#_paste-with-imports-for-javascript-and-typescript) is that this feature should be opt-in, even if it's enabled. In other words, even when the feature is enabled:

1. A regular copy+paste does not copy imports, and the user should expect no performance regression.
2. The user has the option of performing a "paste with imports", either with a different keyboard shortcut, or a different option in the tooltip, which will be slower.

Is this interpretation correct?

**Comment by mjbvz** [comment]
@JavaScriptBach Redacted logs are fine, the main thing I want to see is the timing of each command 

The feature is opt-in in the sense that if you just paste, we will just insert the text and not also add imports. However we still show UI which lets you then select `paste with imports`. That means that still we need to compute if that UI should be shown. This should be relatively fast in most cases but you may be running into an edge case

The `typescript.updateImportsOnPaste.enabled` setting is an escape hatch which you can use to fully disable the feature if it is causing problems

**Comment by JavaScriptBach** [comment]
OK, let me work on redacting my log.

What I'm seeing is that:

1. When I attempt a regular paste, I see a circular spinning icon ranging from several hundred ms to several seconds, and then the text gets pasted without any imports added.
2. I can't see an option in the UI to choose "paste with imports". When I right-click, I just see "paste".

The reason I asked about expected behavior is because I'm wondering whether the performance regression I observed in 1) is expected behavior for large workspaces. If it is, then I would probably have to disable this feature, which would make this bug report moot (for my purposes at least).

**Comment by h0wXD** [comment]
Experiencing the same issue in an angularjs v1 project, which uses dependency injection and doesn't even have any file imports

**Comment by skykep** [comment]
Seems to have fixed the pasting issues, but still takes forever to analyze dependencies.  This became very apparent after upgrading to 1.96.2.

**Comment by brandonscript** [comment]
I've got logs to share of this same issue if it's helpful. Please let me know.

**Comment by charlesfries** [comment]
Off topic comment--frustrating issues like this are the perfect recipe for VSCode going the way of [Atom](https://news.ycombinator.com/item?id=21142934).

**Comment by codekln** [comment]
This issue is really annoying

**Comment by Krechals** [comment]
Adding the following function (which only uses internal types) in a large repo caused the loading spinner to run for 10 seconds, during which the function didn't appear.
```
function foo(fooInput: string, barInput: string) {
  let fooSize = UNDEFINED_VAR;
  if (fooInput) {
    fooSize /= 2;
  }
  if (barInput) {
    fooSize /= 2;
  }
  return fooSize;
}
```

I'm attaching the cpu profile from the linux typescript server:
[paste-trace.cpuprofile](https://github.com/user-attachments/files/18571906/paste-trace.cpuprofile)

System:
- VS Code Version: 1.96.4
- OS Version: Client = Macos 15.2, Remote = Linux 5.4

**Comment by mjbvz** [comment]
The 1.98 VS Code insiders build coming out next week (Feb 10th) adds a workaround that should avoid spinners in the majority of cases. If you're still seeing slowness when using VS Code 1.98, please open a new issue and I'll take a look

You may still see a spinner if you have configured VS Code to try using `paste with imports` by default using the `editor.pasteAs.preferences` setting.  For these cases we still need TS server logs to investigate further. See [my original comment](https://github.com/microsoft/vscode/issues/235959#issuecomment-2539548324) for details on collecting those

---

For current VS Code release (1.97), again you can prevent the spinner from showing on paste by setting: `"typescript.updateImportsOnPaste.enabled": false`, `"javascript.updateImportsOnPaste.enabled": false`. You can also use the `Paste as Text` command to paste just plain text without trying to add any imports

Locking the issue so that this update doesn't get lost. Again if you are still seeing problems on the next 1.98 insiders build, please open up a new issue and I'll take a look
